docs: Add documentation on local cache export (#2035)
dagger/dagger: docs/guides/1230-persistent-cache.md

---
slug: /1230/persistent-cache-with-dagger
displayed_sidebar: 0.2
---

# Persistent cache

> CI that takes eternity is a real pain and can become a bottleneck when your

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

infrastructure and process grow. But Dagger, working with a Buildkit daemon,

> have a powerful cache system that triggers actions only when it's necessary.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

> However, sometime you can't keep the same Buildkit daemon along your CI/CD.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

> For instance, if you use GitHub runner, your daemon will be created on each

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

run and **cache will be lost**.

In this page, we will see how to use `--cache-from` and `--cache-to` flags

> keep a permanent cache, from a local environment to GitHub action.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

## Ephemeral cache

> As an example, we will use a Dagger plan to build a go program.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

> You can use any go project and the following snippet to build it in a

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

standard way

```cue
package ci

import (
  "dagger.io/dagger"
  "universe.dagger.io/go"
)

dagger.#Plan & {
  // Retrieve source from current directory
  // Input
  client: filesystem: ".": read: {
    contents: dagger.#FS

>     include: ["**/*.go", "go.mod", "go.sum"]

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

I added a small comments

  }

  // Output
  client: filesystem: "./output": write: {
    contents: actions.build.output
  }

  actions: {
    // Alias on source
    _source: client.filesystem.".".read.contents

    // Build go binary
    build: go.#Build & {
      source: _source
    }
  }
}
```

To build go binary, just tip `dagger do build`, it should take some time to
install dependencies, build binary and output it.

> 

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

Here's an example of a run

```shell
[✔] actions.build.container                                               338.0s
[✔] client.filesystem.".".read                                              0.1s
[✔] actions.build.container.export                                          0.1s
[✔] client.filesystem."./output".write                                      0.2s
```

Indeed, Dagger has an ephemeral cache so if you rerun it, that shouldn't take
that long.

```shell
[✔] actions.build.container                                                 2.9s
[✔] client.filesystem.".".read                                              0.0s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.1s
```

But if you stop Buildkit daemon and remove volume, cache will be lost and

> all actions will be executed again.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

:::info
Now we have seen how ephemeral cache works, let's continue to understand how

> store cache in your local filesystem, so you can clean your docker engine without

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

losing all your CI's cache.
:::

## Persistent cache in your local filesystem

To store cache in your local filesystem, you don't need much effort : just

> add `--cache-from type=local,mode=max,dest=<output folder>` to `dagger do build`.

Author: I added a tests file in https://github.com/dagger/dagger/pull/2035/commits/18aea2fbbf929671b15a3e9d3e118dea49213d68

But then we should enable tests on documentation.

> add `--cache-from type=local,mode=max,dest=<output folder>` to `dagger do build`.

Reviewer: I personally think it would be fine to just put it in the actual tests dir of our repo. We need tests for cache import/export there anyways.

> add `--cache-from type=local,mode=max,dest=<output folder>` to `dagger do build`.

Author: How would you test it? It's a classic execution but we expect that is faster with cache?
You want me to upload a local cache on our repository and compare time when I mount it and when I do not?

:::tip

> Using `mode=max` argument will cache **all** layers from intermediate

Author: I added a line about `mode=min`, if you think it's not enough maybe we can:
- Add a link to buildkit explanation
- Create a section

> Using `mode=max` argument will cache **all** layers from intermediate

Reviewer: Thanks, Buildkit's explanation is very minimal so I think what we have here is probably good enough for now.

steps, with is really useful in the context of Dagger where you will have
multiple steps to execute.
:::

> Here's an example that exports cache to `storage`.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

```shell
dagger do build --cache-to type=local,mode=max,dest=storage 
# ...

tree storage -L 1   
storage
├── blobs
├── index.json
└── ingest
```

> A new directory has been created that contains cache the run

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

:::caution
If you run different action in the plan, don't forget to store cache in different

> destination, so it will be not overwritten.

Reviewer: I think this is worth explaining in a bit more detail. Basically, our advice is that the directory cache is exported to should be namespaced by the action.

This is also unfortunate because there is no way overlapping cache between the actions can be de-duped. Between that and the performance issues noted by @aluzzardi [here](https://github.com/dagger/dagger/discussions/1875#discussioncomment-2445460), I almost wonder if we should even be emphasizing the use of local cache at all and instead just recommend registry cache. AFAIK, the use case for local cache usually involves exporting to some local dir that is then uploaded to a remote backend and later redownloaded for subsequent CI runs. Given the tradeoffs, I want to be careful we don't end up recommending something to our users that isn't actually beneficial to them.

I don't have all the context that lead to this doc, so if there's a pressing user need for local cache export documentation, then that's totally fine of course, but we should make it clear what performance tradeoffs there are IMO.

Reviewer: Yeah we can still document local exports, but IMO opinion it would be better to start with registry at the top of the doc, use it for explaining the concepts and then mention local export as an option too after. This is subjective though.

I think this `:::caution` note should also be a bit more clear. Something like:
```
Local cache exports will overwrite anything already present in the directory, including any previous cache exports. Cache for distinct actions can be exported to different directories in order to not overwrite each other, but this may currently result in duplicated data if the two actions share any exported cache data in common.
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

:::

To import the cache previously stored, you can use `--cache-from type=local,src=<cache folder>`.

Here's an example, using a new buildkit daemon

```shell
# Down buildkit daemon
docker container stop dagger-buildkitd && docker container rm dagger-buildkitd && docker volume rm dagger-buildkitd

# Import cache on rebuild
dagger do build --cache-to type=local,mode=max,dest=storage --cache-from type=local,src=storage
[✔] actions.build.container                                                 2.3s
[✔] client.filesystem.".".read                                              0.1s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.4s
```

:::info
In this part, we have how to keep cache in a local filesystem, if you want
to see more options on local export, look at [Buildkit cache documentation](https://github.com/moby/buildkit#local-directory-1)
:::

## Persistent cache in a remote registry

> Buildkit can also import/export cache to a registry, it's a great way to share cache between your team and avoid

Reviewer: ```suggestion
Buildkit can also import/export cache to a registry. This is a great way to share cache between your team and avoid
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

flooding your filesystem.

:::tip
Using a registry as cache storage is more efficient than local storage because Buildkit will only re-export
missing layers on multiple runs.
:::

Let's first deploy a simple registry in your localhost

```shell
docker run -d -p 6000:5000 --restart=always --name cache-registry registry:2
```

Then it's not much different from local export

```shell
dagger do build --cache-to type=registry,mode=max,ref=localhost:5000/cache --cache-from type=registry,ref=localhost:5000/cache
[✔] actions.build.container                                                 1.3s
[✔] client.filesystem.".".read                                              0.0s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.1s
```

:::info

> See more options on registry export at [Buildkit cache documentation](https://github.com/moby/buildkit#registry-push-image-and-cache-separately)

Reviewer: When linking to the external buildkit repo, maybe we should use a permalink in case a future update to their README causes our doc's link to no longer work: https://github.com/moby/buildkit/blob/ed5f454d2021c2378229ac68c05cf560cfe25f36/README.md#registry-push-image-and-cache-separately

The downside though is that we need to remember to update the link if Buildkit has a new release w/ updated options here, so not sure, just a thought.

Author: You are right! I think it's better to update the link ourselves that having broken links

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/d916466adfe0a5cb53d851ba7e04033716be0afd

[✔] actions.build.container.export                                          0.1s
[✔] client.filesystem."./output".write                                      0.2s
```

Indeed, Dagger has an ephemeral cache so if you rerun it, that shouldn't take
that long.

```shell
[✔] actions.build.container                                                 2.9s
[✔] client.filesystem.".".read                                              0.0s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.1s
```

But if you stop the Buildkit daemon and remove its storage volume, cache will be lost and
all actions will be executed again.

:::info
Now we have seen how ephemeral cache works, let's continue to understand how
store cache in your local filesystem, so you can clean your Buildkit daemon without
losing all your CI's cache.
:::

## Persistent cache in your local filesystem

To store cache in your local filesystem, you don't need much effort : just
run `dagger do build --cache-from type=local,mode=max,dest=<output folder>`

:::tip
Using `mode=max` argument will cache **all** layers from intermediate

> steps, with is really useful in the context of Dagger where you will have

Reviewer: ```suggestion
steps, which is really useful in the context of Dagger where you will have
```

Author: Nice catch, fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

multiple steps to execute.  

> To only store the final layer, use `mode=min`.

Reviewer: ```suggestion
To only store the final layers of the exported result, use `mode=min`.
```
This is sort of hard to explain briefly but that's my best attempt. I just want to make sure it's clear that it is multiple layers and not just one.

Author: Thank you!
Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

:::

Here's an example that exports cache to a local directory at path `./storage`.

```shell
dagger do build --cache-to type=local,mode=max,dest=storage 
# ...

tree storage -L 1   
storage
├── blobs
├── index.json
└── ingest
```

As shown above, new directory has been created that contains cache artifacts from the run

:::caution
If you run different action in the plan, don't forget to store cache in different
destination, so it will be not overwritten.
:::

To import the cache previously stored, you can use `--cache-from type=local,src=<cache folder>`.

> Here's an example, using a new buildkit daemon

Reviewer: ```suggestion
Here's an example, using a new Buildkit daemon running on a Docker installation
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

```shell
# Down buildkit daemon

> docker container stop dagger-buildkitd && docker container rm dagger-buildkitd && docker volume rm dagger-buildkitd

Reviewer: Should add a warning that this throws all your data away.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

# Import cache on rebuild
dagger do build --cache-to type=local,mode=max,dest=storage --cache-from type=local,src=storage
[✔] actions.build.container                                                 2.3s
[✔] client.filesystem.".".read                                              0.1s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.4s
```

:::info

> In this part, we have how to keep cache in a local filesystem, if you want

Reviewer: ```suggestion
In this part, we have learned how to export and import cache using a local filesystem, if you want
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

Buildkit can also import/export cache to a registry. This is a great way to
share cache between your team and avoid flooding your filesystem.

:::tip
Using a registry as cache storage is more efficient than local storage because Buildkit will only re-export
missing layers on multiple runs.
:::

Let's first deploy a simple registry in your localhost

```shell
docker run -d -p 6000:5000 --restart=always --name cache-registry registry:2
```

Then it's not much different from local export

```shell
dagger do build --cache-to type=registry,mode=max,ref=localhost:5000/cache --cache-from type=registry,ref=localhost:5000/cache
[✔] actions.build.container                                                 1.3s
[✔] client.filesystem.".".read                                              0.0s
[✔] actions.build.container.export                                          0.0s
[✔] client.filesystem."./output".write                                      0.1s
```

:::info
See more options on registry export at [Buildkit cache documentation](https://github.com/moby/buildkit/blob/v0.10.1/README.md#registry-push-image-and-cache-separately)
:::

## Persistent cache in GitHub Actions

> Buildkit has a great integration to store cache from GitHub Action.  

Reviewer: ```suggestion
Buildkit has a builtin support for storing cache with GitHub Action Cache.
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

> That features is really powerful with Dagger because you can cache everything that has not changes in your PR.

Reviewer: ```suggestion
```

Probably just drop this line, this should be clear already IMO.

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

> It's not much different as `local` or `registry` exports, let's integrate cache in a
> simple workflow

Reviewer: ```suggestion
This cache backend is not that different from `local` or `registry` exports. Let's integrate cache in a
simple workflow using the following workflow:
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

I replaced the second `workflow` with `the following one` to avoid repetition

```yaml title=".github/workflows/build-example.cue"

> name: "Dagger export export"

Reviewer: ```suggestion
name: "Dagger Export"
```

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

on:
  push:
    branches: [ main ]

jobs:
  build:
    name: "Build"
    runs-on: ubuntu-latest
    # Set cache export environment
    env:
      DAGGER_CACHE_FROM: type=gha,scope=dagger-cache-example
      DAGGER_CACHE_TO: type=gha,mode=max,scope=dagger-cache-example
    steps:
      - name: Checkout
        uses: actions/checkout@v2
          
      # Required to retrieve GitHub token    
      - name: Expose GitHub Runtime

>         uses: crazy-max/ghaction-github-runtime@v1

Reviewer 2: Not required. See https://github.com/dagger/dagger-for-github/pull/56#issuecomment-1111730526

Author: Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66

all actions will be executed again.

:::info
Now we have seen how ephemeral cache works, let's continue to understand how
store cache in your local filesystem, so you can clean your Buildkit daemon without
losing all your CI's cache.
:::

## Persistent cache in a remote registry

Buildkit can also import/export cache to a registry. This is a great way to
share cache between your team and avoid flooding your filesystem.

:::tip
Using a registry as cache storage is more efficient than local storage because Buildkit will only re-export
missing layers on multiple runs.
:::

To store cache in an external registry, you just need to add flags to dagger command :
`dagger do <action> --cache-to type=registry,mode=max,ref=<registry target>/<image> --cache-from type=registry,ref=<registry target>/<image>`

:::tip
Using `mode=max` argument will cache **all** layers from intermediate
steps, which is really useful in the context of Dagger where you will have
multiple steps to execute.  
To only store the final layers of the exported result, use `mode=min`.
:::

Let's first deploy a simple registry in your localhost

```shell

> docker run -d -p 6000:5000 --restart=always --name cache-registry registry:2

Reviewer 3: ```suggestion
docker run -d -p 5000:5000 --restart=always --name cache-registry registry:2
```

There is a port mismatch between this line and line 121.
Either this value needs to be changed to 5000 or line 121 needs to change to 6000

Reviewer 4: Resolved in https://github.com/dagger/dagger/pull/2035/commits/ebac13380004ecaa71f6a37dcd68e02ffad71943. Thanks @tdjordan!
